### ISSUE TITLE
torch.nn.Bilinear weights initialization reason

### ISSUE THREAD

**Issue by redwrasse** [pytorch/pytorch#132231]
### 📚 The doc issue

`torch.nn.Bilinear` initializes weights uniformly in +-1/in-1-features (https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/linear.py#L219) , and this fact is included in the documentation.

Is there a justification for this weights initialization? A one-line code comment would help. Intuitively, would expect instead that given a bilinear form `x^TAy`, weights initialization would be symmetric under `x <-> y`. 

```
    def reset_parameters(self) -> None:
        # FIXME: weights are initialized to blah blah because of blah blah
        bound = 1 / math.sqrt(self.weight.size(1))
        init.uniform_(self.weight, -bound, bound)
        if self.bias is not None:
            init.uniform_(self.bias, -bound, bound)
```

**Comment by albanD** [comment]
My best guess is to make it similar to the Linear layer behavior.

**Comment by redwrasse** [comment]
As an expression it's similar, as a quantity not necessarily. Say x dim is an order of magnitude larger than y dim. I would expect symmetry under `x <-> y` (so maybe take average), or call this out as a convention where in typical use cases x is such and such and y is such and such.

**Comment by redwrasse** [comment]
One approach might be, following the logic of eq. (13) of the Glorot initialization paper, try to find an initialization that guarantees `n * var(W)` is 1, or is the same value as in the sister Linear module `n * var(W) = 1/3` there.

This in term would depend on what x, y represent in the bilinear form x^TAy. I'm assuming x, y most naturally would be partitions of the input, say x is left partition and y right partition of the input (so input = x (direct sum) y). Then, work through the math to see what the constraint on var(W) should be.

**Comment by redwrasse** [comment]
Doing some napkin math (needs to be double-checked) where x, y in `x^TWy` are assumed partitions of the input, and following https://proceedings.mlr.press/v9/glorot10a/glorot10a.pdf, I get 
`var(W) = 2 / (x_dim * y_dim)` to prevent any growth in gradients wrt earlier layers, so order of magnitude translates to uniform initialization `Unif[-sqrt(6)/sqrt(x_dim*y_dim), sqrt(6)/sqrt(x_dim*y_dim)]`, or if matching the current linear module implementation with the missing sqrt(3) factor, `Unif[-sqrt(2)/sqrt(x_dim*y_dim), sqrt(2)/sqrt(x_dim*y_dim)]`.

**Comment by redwrasse** [comment]
To get started, it may help to dig up the original intended/current intended use case for Bilinear layers.
